<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>

Юлия, доброго дня! рад приветствовать тебя вновь) <br />
(А дальше - шаблонная часть, как ты заметила по другим спринтам)<br>
<br />
Меня зовут Николай Шавлюго. <br />И на этом этапе твоего движения к новой профессии от меня будут комментарии по написанному коду. <br />Чтобы меньше "мусорить эфир" и чтобы максимально наглядно отделяться от основного кода, есть предложение условиться в некоторых обозначениях:<br /> <br />
<u><b>ТАКОЙ ШРИФТ</b></u> - всегда начало комментария <br />
<font color='green'>такой шрифт</font> - комментарии о том, что всё ОК <br/>
<font color='orange'>такой шрифт</font> - комментарии о том, что всё ОК по результату,<br> однако есть на что обратить внимание в плане применения техник, или есть способы сделать более короткий или быстрый код <br/>
<font color='red'>такой шрифт</font> - комментарии о том, что есть критичный момент, влияющий на бизнес-результат проекта.<br/>
</font><br /><br />
<font color='green'>Моей целью является не "уличить" в не знании, а просто высказать сверху твоих знаний - свой опыт, что бы ты мог использовать его для своего дальнейшего успеха) И очень здорово будет, если тебе удастся задавать вопросы, да и вообще - всячески доставать меня, если я по каким-то причинам не приму проект:) При этом, из своего опыта скажу, хорошо и важно, когда переписка ревьюера и студента - сохраняется на следующие проекты и даже на будущую практическую деятельность.<br>
<BR> В ПУТЬ!<br>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## Загрузка данных

In [8]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [9]:
df = pd.read_csv('/datasets/insurance.csv')

In [10]:
df.head()

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
Пол                  5000 non-null int64
Возраст              5000 non-null float64
Зарплата             5000 non-null float64
Члены семьи          5000 non-null int64
Страховые выплаты    5000 non-null int64
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


In [12]:
df.describe().T

count        mean          std     min      25%      50%  \
Пол                5000.0      0.4990     0.500049     0.0      0.0      0.0   
Возраст            5000.0     30.9528     8.440807    18.0     24.0     30.0   
Зарплата           5000.0  39916.3600  9900.083569  5300.0  33300.0  40200.0   
Члены семьи        5000.0      1.1942     1.091387     0.0      0.0      1.0   
Страховые выплаты  5000.0      0.1480     0.463183     0.0      0.0      0.0   

                       75%      max  
Пол                    1.0      1.0  
Возраст               37.0     65.0  
Зарплата           46600.0  79000.0  
Члены семьи            2.0      6.0  
Страховые выплаты      0.0      5.0

In [13]:
df.duplicated().sum()

153

In [14]:
df = df.drop_duplicates().reset_index(drop = True)

<b>Вывод</b>

Данные содержат 5000 строк и 5 столбцов. Аномальных значений не обнаружено, но выявлено 153 дубликата, которые принято решение удалить.

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>
</font>
<font color='green'><br>
Данные на базе) ОК, хорошо, когда есть аналитическая культура - обязательно делать первичный анализ данных, не смотря на проект)

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Ответ:** не изменится.

**Обоснование:**  

Введем обозначения:  

$
P = X^{-1}
$  

$
a' = Xw'
$  

$
w' = ((XP)^T XP)^{-1}(XP)^Ty
$  

$
a' = XP((XP)^T XP)^{-1}(XP)^Ty = XP(P^TX^TXP)^{-1}P^TX^Ty = XP(X^TXP)^{-1}(P^T)^{-1}P^TX^Ty
$  

$
a' = XPP^{-1}(X^TX)^{-1}(P^T)^{-1}P^TX^Ty = XE(X^TX)^{-1}EX^Ty = X(X^TX)^{-1}X^Ty = Xw
$


<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
<font color='green'> 
    <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА 2</u></b>
</font>
<font color='green'><br>ок, есть контакт)

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>
</font>
<font color='red'><br>
оооочень лаконично  получилось. Поэтому, не смог уловить переходы, извини.... Может поподробнее удастся расписать? <br>
    
А ещё есть такой способ: через выражение (при подстановке $w' = ((XP)^T XP)^{-1} (XP)^T y$) надо дойти до $w' = P^{-1}w$ , где $w = (X^T X)^{-1} X^T y$ Это и будет доказательством.

## Алгоритм преобразования

**Алгоритм**

Персональная информация представлена ввиде набора данных.  
Для начала представим датафрейм ввиде матрицы. Далее для того, чтобы зашифровать данные клиентов, исходную матрицу преобразуем в обратную, а предсказания получим путем умножения обратной матрицы на вектор весов линейной регрессии.

**Обоснование**

Как было доказано выше, качество линейной регрессии не изменится при умножении признаков на обратную матрицу, но данные клиентов останутся зашифроваными.

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>
</font>
<font color='green'><br>
Тут как раз и нужен план действий для пункта 4) Делай раз, делай два, делай три..Например так:<br>
1) создаём рандомную матрицу<br>
2) проверяем её на обратимость (создаём единичную матрицу)<br>
3) Создаём новую матрицу данных<br>
4) Ищем метрики качества на исходных данных и на преобразованных<br>
5) делаем вывод

## Проверка алгоритма

<b>Разделение данных на выборки</b>

In [15]:
features = df.drop('Страховые выплаты', axis=1)
target = df['Страховые выплаты']

<b>Модель Линейная регрессия из sklern</b>

In [16]:
model = LinearRegression()
model.fit(features, target)
prediction = model.predict(features)
print(r2_score(target, prediction))

0.4302010044852067


<b>Примение алгоритма шифрования пользовательских данных</b>

In [17]:
class LinearRegressionNew:
    def fit(self, train_features, train_target):
        X = np.concatenate((np.ones((train_features.shape[0], 1)), train_features), axis=1)
        y = train_target
        w = ((np.linalg.inv(X.T @ X) @ X.T) @ y)
        self.w = w[1:]
        self.w0 = w[0]

    def predict(self, test_features):
        return test_features.dot(self.w) + self.w0
    
model_new = LinearRegressionNew()
model_new.fit(features, target)
predictions = model_new.predict(features)
print(r2_score(target, predictions))

0.4302010044852068


<b>Вывод</b>

В результате проделанной работы осуществлена предварительная подготовка данных: из выборки удалены дубликаты. Далее разработан алгоритм шифрования данных клиентов, условие использования которого не должно ухудшить качетсва модели. Для проверки предположения о том, что использование алгоритма не повлечет потерю качества модели, была получена метрика R2 путем применения линейной регрессии из sklearn, затем создан класс для непосредственной реализации модифицированной линейной регрессиии. Метрка качества R2 преобразованной модели оказалась равной метрики качества R2 модели из sklearn.

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>
</font>
<font color='green'><br>
ОК, Юлия, и с линейной алгеброй ты справился)<br>
Всё, в общем-то, по делу, со смыслом. Доказательство в п.2 - теперь понятно)<br>
Да и весь проект - на уровне.<br>
    
Если рассматривать передачу данных третьим лицам, то к преимуществам данного подхода можно отнести, в первую очередь, ощутимую экономию ресурсов по сравнению с некоторыми известными алгоритмами шифрования, в особенности это касается ассиметричных алгоритмов с достаточно длинными ключами (AES, RSA), которые требуют много процессорного времени.<br><br>
    
К главному недостатку подхода можно отнести увеличение сложности проверки на вырожденность преобразования с ростом количества признаков в выборке. В рассмотренном случае необходимо подобрать подходящую матрицу размера  4×4 . Однако, если признаки будут исчисляться тысячами, что в теории всегда надо положить, то расчет соответствующих определителей может быть крайне ресурсозатратным. Но и эту проблему можно будет решить)<br><br>
<font color='green'><font color='green'>
Спасибо) Успешного дальнейшего обучения

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>
</font>
<font color='green'><br>
Юлия, отличная работа! Сама практическая цель достигнута, ОК<br>
Тут всё верно. Есть всего один момент для обсуждения:<br>
    
В разделе "Умножение матриц" оставил вопрос, посмотри, пожалуйста.<font color='red'><br>
<br><font color='green'><font color='green'>
Спасибо) Буду ждать.

## Чек-лист проверки

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные загружены
- [x]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [x]  Указан правильный вариант ответа
    - [x]  Вариант обоснован
- [x]  Выполнен шаг 3: предложен алгоритм преобразования
    - [x]  Алгоритм описан
    - [x]  Алгоритм обоснован
- [x]  Выполнен шаг 4: алгоритм проверен
    - [x]  Алгоритм реализован
    - [x]  Проведено сравнение качества моделей до и после преобразования